modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/pipelining/HttpPipeliningHandler.java

```diff
@@ -80,14 +78,19 @@ public void write(final ChannelHandlerContext ctx, final Object msg, final Chann
                     while (!holdingQueue.isEmpty()) {
                         /*
                          * Since the response with the lowest sequence number is the top of the priority queue, we know if its sequence
-                         * number does not match the current write sequence then we have not processed all preceding responses yet.
+                         * number does not match the current write sequence number then we have not processed all preceding responses yet.
                          */
                         final HttpPipelinedResponse response = holdingQueue.peek();
                         if (response.sequence() != writeSequence) {
                             break;
                         }
                         holdingQueue.remove();
-                        ctx.write(response.response(), promise);
+                        /*
+                         * We must use the promise attached to the response; this is necessary since are going to hold a response until all
+                         * responses that precede it in the pipeline are written first. Note that the promise from the method invocation is
+                         * not ignored, it will already be attached to an existing response and consumed when that response is drained.
+                         */
+                        ctx.write(response.response(), response.promise());
```

Comment:
Should we assert at the beginning of this if clause that the promise in the HttpPipelinedResponse is that one we got as a parameter?